deployments/apiv2/services/users/src/main.py

```diff
@@ -0,0 +1,131 @@
+from models.users import UserLogin, UserCreate, UserProfile
+from models.token import JWTMeta, JWTDetails, JWTPayload, AccessToken
+from models.errors import LoginError, RegistrationError
+
+db = Database()
+app = FastAPI()
+
+
+@app.middleware("http")
+async def db_session_middleware(request: Request, call_next):
+    request.state.pgpool = db.pool
+    response = await call_next(request)
+    return response
+
+
+@app.on_event("startup")
+async def startup():
+    await db.create_pool()
+
+
+@app.on_event("shutdown")
+async def shutdown():
+    await db.close()
+
+
+@app.get("/users/me", response_model=UserProfile)
+async def index(request: Request, token=Depends(ProtectedAny(scope=["users:free"]))):
+    try:
+        async with request.state.pgpool.acquire() as con:
+            user_id = uuid.UUID(hex=token["userid"])
+            rows = await con.fetchrow("select name, email, user_id, account_type, created_at, updated_at, data->'scope' as scope from users where user_id = $1", user_id)
+
+            return UserProfile(
+                username=rows.get("name", ""),
+                email=rows.get("email", ""),
+                user_id=rows.get("user_id","").hex if "user_id" in rows else "",
+                account_type=rows.get("account_type", ""),
+                scope=json.loads(rows.get("scope", "[]")),
+            )
+    except:
+        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+
+@app.post("/login", response_model=AccessToken)
+async def login(request: Request, details: UserLogin):
+    ph = PasswordHasher()
+    failed_msg = "Username and/or password is incorrect"
+
+    try:
+        async with request.state.pgpool.acquire() as con:
+            query = "SELECT password, user_id, data->'scope' as scope FROM users where deleted = 'f' and name = $1"
+            row = await con.fetchrow(query, details.username)
+
+            try:
+                ph.verify(row.get("password", "x"*100), details.password.get_secret_value())
+            except VerifyMismatchError:
+                raise LoginError(failed_msg)
+            except InvalidHash:
+                """
+                  the user wasn't found but we don't want to leak info about valid users
+                  through timing analysis so we still hash the supplied password before
+                  returning an error
+                """
+                ph.hash(p)
+                raise LoginError(failed_msg)
+            else:
+                scope = json.loads(row.get("scope", "[]"))
+                iat = timegm(datetime.now(tz=timezone.utc).utctimetuple())
+                exp = timegm((datetime.now(tz=timezone.utc) + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)).utctimetuple())
+
+                jwt_meta = JWTMeta(aud=JWT_AUDIENCE, scope=scope, iat=iat, exp=exp)
+                jwt_details = JWTDetails(userid=row["user_id"].hex)
+                jwt_payload = JWTPayload(**jwt_meta.dict(), **jwt_details.dict())
```

Comment:
also, is the ** in python the same as the spread operator in js -> [...array] ?

Reply:
yeah pretty much, `**` splats `key=value` for each item in a dictionary, for any array in python you would use a single `*`

you'll see stuff like

```python
def some_func(*args, **kwargs):
    ....
```
in python that spreads the unamed arguments from a list and then all the key word arguments from a dict